File: azure-functions/src/services/userService.ts
import { getConnection, sql } from '../config/database';
import { User, CreateUserRequest, JwtPayload, TokenPair } from '../types';
import { hashPassword, verifyPassword } from '../utils/password';
import { generateTokens } from '../utils/jwt';
import { config } from '../config/env';

export class UserService {
  /**
   * Create a new user with default password
   */
  async createUserWithDefaultPassword(data: CreateUserRequest): Promise<{ user: User; tokens: TokenPair; defaultPassword: string }> {
    const pool = await getConnection();
    const transaction = pool.transaction();

    try {
      await transaction.begin();

      // Check if email already exists
      const emailCheck = await transaction.request()
        .input('Email', sql.NVarChar(256), data.email)
        .query(`
          SELECT UserId FROM [auth].[Users]
          WHERE Email = @Email AND DeletedAtUtc IS NULL
        `);

      if (emailCheck.recordset.length > 0) {
        throw new Error('User with this email already exists');
      }

      // Verify company exists
      const companyCheck = await transaction.request()
        .input('CompanyId', sql.BigInt, data.companyId)
        .query(`
          SELECT CompanyId, IsActive FROM [core].[Companies]
          WHERE CompanyId = @CompanyId AND DeletedAtUtc IS NULL
        `);

      if (companyCheck.recordset.length === 0) {
        throw new Error('Company not found');
      }

      if (!companyCheck.recordset[0].IsActive) {
        throw new Error('Company is not active');
      }

      // Hash default password
      const defaultPassword = config.app.defaultPassword;
      const passwordHash = await hashPassword(defaultPassword);

      // Insert user
      const userResult = await transaction.request()
        .input('CompanyId', sql.BigInt, data.companyId)
        .input('Email', sql.NVarChar(256), data.email)
        .input('DisplayName', sql.NVarChar(128), data.displayName)
        .input('PasswordHash', sql.VarBinary(256), passwordHash)
        .query(`
          INSERT INTO [auth].[Users] (
            CompanyId, Email, DisplayName, PasswordHash, IsActive
          )
          OUTPUT INSERTED.*
          VALUES (
            @CompanyId, @Email, @DisplayName, @PasswordHash, 1
          )
        `);

      const user = userResult.recordset[0] as User;

      // Assign role (default to COMPANY_ADMIN if not specified)
      const roleId = data.roleId || 2; // 2 = COMPANY_ADMIN
      await transaction.request()
        .input('UserId', sql.BigInt, user.UserId)
        .input('RoleId', sql.Int, roleId)
        .query(`
          INSERT INTO [auth].[UserRoles] (UserId, RoleId)
          VALUES (@UserId, @RoleId)
        `);

      // Get user roles
      const rolesResult = await transaction.request()
        .input('UserId', sql.BigInt, user.UserId)
        .query(`
          SELECT r.Name
          FROM [auth].[UserRoles] ur
          JOIN [auth].[Roles] r ON r.RoleId = ur.RoleId
          WHERE ur.UserId = @UserId
        `);

      const roles = rolesResult.recordset.map((r: any) => r.Name);

      await transaction.commit();

      // Generate tokens
      const jwtPayload: JwtPayload = {
        userId: user.UserId,
        email: user.Email,
        companyId: user.CompanyId,
        roles,
      };

      const tokens = generateTokens(jwtPayload);

      return { user, tokens, defaultPassword };
    } catch (error) {
      await transaction.rollback();
      throw error;
    }
  }

  /**
   * Login user
   */
  async login(email: string, password: string, ipAddress?: string, userAgent?: string): Promise<{ user: User; tokens: TokenPair }> {
    const pool = await getConnection();

    try {
      // Get user with roles
      const userResult = await pool.request()
        .input('Email', sql.NVarChar(256), email)
        .query(`
          SELECT u.*, r.Name as RoleName
          FROM [auth].[Users] u
          LEFT JOIN [auth].[UserRoles] ur ON ur.UserId = u.UserId
          LEFT JOIN [auth].[Roles] r ON r.RoleId = ur.RoleId
          WHERE u.Email = @Email AND u.DeletedAtUtc IS NULL
        `);

      if (userResult.recordset.length === 0) {
        // Log failed attempt
        await this.logLoginAttempt(email, null, false, 'User not found', ipAddress, userAgent);
        throw new Error('Invalid email or password');
      }

      const user = userResult.recordset[0] as User;
      const roles = userResult.recordset.map((r: any) => r.RoleName).filter((r: string) => r);

      // Check if user is active
      if (!user.IsActive) {
        await this.logLoginAttempt(email, user.UserId, false, 'User is inactive', ipAddress, userAgent);
        throw new Error('User account is inactive');
      }

      // Check if user is locked
      if (user.IsLocked) {
        await this.logLoginAttempt(email, user.UserId, false, 'User is locked', ipAddress, userAgent);
        throw new Error('User account is locked due to too many failed login attempts');
      }

      // Verify password
      const isValidPassword = await verifyPassword(password, user.PasswordHash);

      if (!isValidPassword) {
        // Increment failed login attempts
        await this.incrementFailedLoginAttempts(user.UserId);
        await this.logLoginAttempt(email, user.UserId, false, 'Invalid password', ipAddress, userAgent);
        throw new Error('Invalid email or password');
      }

      // Reset failed login attempts
      await this.resetFailedLoginAttempts(user.UserId);

      // Update last login
      await pool.request()
        .input('UserId', sql.BigInt, user.UserId)
        .query(`
          UPDATE [auth].[Users]
          SET LastLoginUtc = SYSUTCDATETIME()
          WHERE UserId = @UserId
        `);

      // Log successful login
      await this.logLoginAttempt(email, user.UserId, true, null, ipAddress, userAgent);

      // Generate tokens
      const jwtPayload: JwtPayload = {
        userId: user.UserId,
        email: user.Email,
        companyId: user.CompanyId,
        roles,
      };

      const tokens = generateTokens(jwtPayload);

      return { user, tokens };
    } catch (error) {
      throw error;
    }
  }

  /**
   * Set/Reset user password
   */
  async setPassword(userId: number, newPassword: string): Promise<void> {
    const pool = await getConnection();

    try {
      // Get user
      const userResult = await pool.request()
        .input('UserId', sql.BigInt, userId)
        .query(`
          SELECT UserId, IsActive FROM [auth].[Users]
          WHERE UserId = @UserId AND DeletedAtUtc IS NULL
        `);

      if (userResult.recordset.length === 0) {
        throw new Error('User not found');
      }

      const user = userResult.recordset[0];

      if (!user.IsActive) {
        throw new Error('User is not active');
      }

      // Hash new password
      const passwordHash = await hashPassword(newPassword);

      // Update password and reset failed attempts if locked
      await pool.request()
        .input('UserId', sql.BigInt, userId)
        .input('PasswordHash', sql.VarBinary(256), passwordHash)
        .query(`
          UPDATE [auth].[Users]
          SET PasswordHash = @PasswordHash,
              FailedLoginAttempts = 0,
              IsLocked = 0,
              LastFailedLoginUtc = NULL
          WHERE UserId = @UserId
        `);
    } catch (error) {
      throw error;
    }
  }

  /**
   * Get user by ID
   */
  async getUserById(userId: number): Promise<User | null> {
    const pool = await getConnection();
    const result = await pool.request()
      .input('UserId', sql.BigInt, userId)
      .query(`
        SELECT * FROM [auth].[Users]
        WHERE UserId = @UserId AND DeletedAtUtc IS NULL
      `);

    return result.recordset[0] as User || null;
  }

  /**
   * Increment failed login attempts
   */
  private async incrementFailedLoginAttempts(userId: number): Promise<void> {
    const pool = await getConnection();
    await pool.request()
      .input('UserId', sql.BigInt, userId)
      .query(`
        UPDATE [auth].[Users]
        SET FailedLoginAttempts = FailedLoginAttempts + 1,
            LastFailedLoginUtc = SYSUTCDATETIME(),
            IsLocked = CASE WHEN FailedLoginAttempts >= 4 THEN 1 ELSE 0 END
        WHERE UserId = @UserId
      `);
  }

  /**
   * Reset failed login attempts
   */
  private async resetFailedLoginAttempts(userId: number): Promise<void> {
    const pool = await getConnection();
    await pool.request()
      .input('UserId', sql.BigInt, userId)
      .query(`
        UPDATE [auth].[Users]
        SET FailedLoginAttempts = 0,
            LastFailedLoginUtc = NULL,
            IsLocked = 0
        WHERE UserId = @UserId
      `);
  }

  /**
   * Log login attempt
   */
  private async logLoginAttempt(
    email: string,
    userId: number | null,
    success: boolean,
    failureReason: string | null,
    ipAddress?: string,
    userAgent?: string
  ): Promise<void> {
    const pool = await getConnection();
    await pool.request()
      .input('Email', sql.NVarChar(256), email)
      .input('UserId', sql.BigInt, userId)
      .input('IpAddress', sql.VarChar(45), ipAddress || null)
      .input('UserAgent', sql.NVarChar(512), userAgent || null)
      .input('Success', sql.Bit, success)
      .input('FailureReason', sql.NVarChar(256), failureReason)
      .query(`
        INSERT INTO [auth].[LoginAttempts] (Email, UserId, IpAddress, UserAgent, Success, FailureReason)
        VALUES (@Email, @UserId, @IpAddress, @UserAgent, @Success, @FailureReason)
      `);
  }
}

File: azure-functions-csharp/Services/UserService.cs
using Dapper;
using KamanAzureFunctions.DTOs;
using KamanAzureFunctions.Helpers;
using KamanAzureFunctions.Models;
using System.Data;

namespace KamanAzureFunctions.Services;

public class UserService
{
    private readonly DatabaseHelper _dbHelper;
    private readonly JwtHelper _jwtHelper;
    private readonly string _defaultPassword;

    public UserService(DatabaseHelper dbHelper, JwtHelper jwtHelper, string defaultPassword)
    {
        _dbHelper = dbHelper;
        _jwtHelper = jwtHelper;
        _defaultPassword = defaultPassword;
    }

    /// <summary>
    /// Create a new user with default password
    /// </summary>
    public async Task<(User User, AuthenticationResponse Tokens, string DefaultPassword)> CreateUserWithDefaultPasswordAsync(CreateUserRequest request)
    {
        using var connection = _dbHelper.GetConnection();
        await connection.OpenAsync();

        using var transaction = connection.BeginTransaction();

        try
        {
            // Check if email already exists
            var existingUser = await connection.QueryFirstOrDefaultAsync<User>(
                "[auth].[sp_CheckUserEmailExists]",
                new { Email = request.Email },
                transaction,
                commandType: CommandType.StoredProcedure
            );

            if (existingUser != null)
            {
                throw new InvalidOperationException("User with this email already exists");
            }

            // Verify company exists and is active
            var company = await connection.QueryFirstOrDefaultAsync<Company>(
                "[core].[sp_GetCompanyById]",
                new { CompanyId = request.CompanyId },
                transaction,
                commandType: CommandType.StoredProcedure
            );

            if (company == null)
            {
                throw new InvalidOperationException("Company not found");
            }

            if (!company.IsActive)
            {
                throw new InvalidOperationException("Company is not active");
            }

            // Hash default password
            var passwordHash = PasswordHelper.HashPassword(_defaultPassword);

            // Insert user
            var userIdResult = await connection.QuerySingleAsync<dynamic>(
                "[auth].[sp_InsertUser]",
                new
                {
                    CompanyId = request.CompanyId,
                    Email = request.Email,
                    DisplayName = request.DisplayName,
                    PasswordHash = passwordHash,
                    IsActive = true
                },
                transaction,
                commandType: CommandType.StoredProcedure
            );

            long userId = userIdResult.UserId;

            // Retrieve the inserted user
            var user = await connection.QuerySingleAsync<User>(
                "[auth].[sp_GetUserById]",
                new { UserId = userId },
                transaction,
                commandType: CommandType.StoredProcedure
            );

            // Assign role (default to COMPANY_ADMIN if not specified)
            var roleId = request.RoleId ?? 2; // 2 = COMPANY_ADMIN
            await connection.ExecuteAsync(
                "[auth].[sp_AssignRoleToUser]",
                new { UserId = userId, RoleId = roleId },
                transaction,
                commandType: CommandType.StoredProcedure
            );

            // Get user roles
            var roles = await connection.QueryAsync<string>(
                "[auth].[sp_GetUserRoles]",
                new { UserId = userId },
                transaction,
                commandType: CommandType.StoredProcedure
            );

            transaction.Commit();

            // Generate tokens
            var jwtPayload = new UserTokenClaims
            {
                UserId = user.UserId,
                Email = user.Email,
                CompanyId = user.CompanyId,
                Roles = roles.ToList()
            };

            var tokens = _jwtHelper.GenerateTokens(jwtPayload);

            return (user, tokens, _defaultPassword);
        }
        catch
        {
            transaction.Rollback();
            throw;
        }
    }

    /// <summary>
    /// Login user
    /// </summary>
    public async Task<(User User, AuthenticationResponse Tokens)> LoginAsync(string email, string password, string? ipAddress = null, string? userAgent = null)
    {
        using var connection = _dbHelper.GetConnection();

        // Get user for login
        var user = await connection.QueryFirstOrDefaultAsync<User>(
            "[auth].[sp_GetUserByEmailForLogin]",
            new { Email = email },
            commandType: CommandType.StoredProcedure
        );

        if (user == null)
        {
            await LogLoginAttemptAsync(email, null, false, "User not found", ipAddress, userAgent);
            throw new UnauthorizedAccessException("Invalid email or password");
        }

        // Get user roles
        var roles = (await connection.QueryAsync<string>(
            "[auth].[sp_GetUserRoles]",
            new { UserId = user.UserId },
            commandType: CommandType.StoredProcedure
        )).ToList();

        // Check if user is active
        if (!user.IsActive)
        {
            await LogLoginAttemptAsync(email, user.UserId, false, "User is inactive", ipAddress, userAgent);
            throw new UnauthorizedAccessException("User account is inactive");
        }

        // Check if user is locked
        if (user.IsLocked)
        {
            await LogLoginAttemptAsync(email, user.UserId, false, "User is locked", ipAddress, userAgent);
            throw new UnauthorizedAccessException("User account is locked due to too many failed login attempts");
        }

        // Verify password
        var isValidPassword = PasswordHelper.VerifyPassword(password, user.PasswordHash);

        if (!isValidPassword)
        {
            await IncrementFailedLoginAttemptsAsync(user.UserId);
            await LogLoginAttemptAsync(email, user.UserId, false, "Invalid password", ipAddress, userAgent);
            throw new UnauthorizedAccessException("Invalid email or password");
        }

        // Update last login and reset failed attempts
        await connection.ExecuteAsync(
            "[auth].[sp_UpdateLastLogin]",
            new { UserId = user.UserId, IpAddress = ipAddress, UserAgent = userAgent },
            commandType: CommandType.StoredProcedure
        );

        // Log successful login
        await LogLoginAttemptAsync(email, user.UserId, true, null, ipAddress, userAgent);

        // Generate tokens
        var jwtPayload = new UserTokenClaims
        {
            UserId = user.UserId,
            Email = user.Email,
            CompanyId = user.CompanyId,
            Roles = roles
        };

        var tokens = _jwtHelper.GenerateTokens(jwtPayload);

        return (user, tokens);
    }

    /// <summary>
    /// Set/Reset user password
    /// </summary>
    public async Task SetPasswordAsync(long userId, string newPassword)
    {
        using var connection = _dbHelper.GetConnection();

        // Get user to validate
        var user = await connection.QueryFirstOrDefaultAsync<User>(
            "[auth].[sp_GetUserById]",
            new { UserId = userId },
            commandType: CommandType.StoredProcedure
        );

        if (user == null)
        {
            throw new InvalidOperationException("User not found");
        }

        if (!user.IsActive)
        {
            throw new InvalidOperationException("User is not active");
        }

        // Hash new password
        var passwordHash = PasswordHelper.HashPassword(newPassword);

        // Update password and unlock/reset failed attempts
        await connection.ExecuteAsync(
            "[auth].[sp_ResetPasswordAndUnlock]",
            new { UserId = userId, PasswordHash = passwordHash },
            commandType: CommandType.StoredProcedure
        );
    }

    /// <summary>
    /// Get user by ID
    /// </summary>
    public async Task<User?> GetUserByIdAsync(long userId)
    {
        using var connection = _dbHelper.GetConnection();
        return await connection.QueryFirstOrDefaultAsync<User>(
            "[auth].[sp_GetUserById]",
            new { UserId = userId },
            commandType: CommandType.StoredProcedure
        );
    }

    /// <summary>
    /// Get user with roles
    /// </summary>
    public async Task<(User User, List<string> Roles)> GetUserWithRolesAsync(long userId)
    {
        using var connection = _dbHelper.GetConnection();

        // Call stored procedure that returns multiple result sets
        using var multi = await connection.QueryMultipleAsync(
            "[auth].[sp_GetUserWithRoles]",
            new { UserId = userId },
            commandType: CommandType.StoredProcedure
        );

        var user = await multi.ReadSingleOrDefaultAsync<User>();
        if (user == null)
        {
            throw new InvalidOperationException("User not found");
        }

        var roles = (await multi.ReadAsync<string>()).ToList();

        return (user, roles);
    }

    /// <summary>
    /// Increment failed login attempts
    /// </summary>
    private async Task IncrementFailedLoginAttemptsAsync(long userId)
    {
        using var connection = _dbHelper.GetConnection();
        await connection.ExecuteAsync(
            "[auth].[sp_TrackFailedLogin]",
            new { UserId = userId, IpAddress = (string?)null, UserAgent = (string?)null },
            commandType: CommandType.StoredProcedure
        );
    }

    /// <summary>
    /// Reset failed login attempts
    /// </summary>
    private async Task ResetFailedLoginAttemptsAsync(long userId)
    {
        using var connection = _dbHelper.GetConnection();
        await connection.ExecuteAsync(
            "[auth].[sp_UpdateLastLogin]",
            new { UserId = userId, IpAddress = (string?)null, UserAgent = (string?)null },
            commandType: CommandType.StoredProcedure
        );
    }

    /// <summary>
    /// Log login attempt
    /// </summary>
    private async Task LogLoginAttemptAsync(
        string email,
        long? userId,
        bool success,
        string? failureReason,
        string? ipAddress,
        string? userAgent)
    {
        // Keep inline SQL for logging (optional table, not critical)
        using var connection = _dbHelper.GetConnection();

        try
        {
            await connection.ExecuteAsync(
                @"INSERT INTO [auth].[LoginAttempts]
                  (Email, UserId, IpAddress, UserAgent, Success, FailureReason)
                  VALUES (@Email, @UserId, @IpAddress, @UserAgent, @Success, @FailureReason)",
                new
                {
                    Email = email,
                    UserId = userId,
                    IpAddress = ipAddress,
                    UserAgent = userAgent,
                    Success = success,
                    FailureReason = failureReason
                }
            );
        }
        catch
        {
            // Ignore logging errors - don't fail the operation
        }
    }
}
